design/ceph/object/ceph-bucket-notification-crd.md

```diff
@@ -0,0 +1,91 @@
+# Ceph bucket notifications CRD
+
+## Overview
+
+Ceph added support for the bucket notifications feature from Nautilus onwards. It allows sending messages to various endpoints when a new event occurs on a bucket [ref](https://docs.ceph.com/docs/master/radosgw/notifications/)
+
+Setup of those notifications are normally done by sending HTTP requests to the RGW, either to create/delete topics pointing to specific endpoints, or create/delete bucket notifications based on those topics.
+
+This functionality eases this process by avoiding to use external tools or scripts. It is replaced by creation of CR definitions that contain all the information necessary to create topics and/or notifications, which the rook operator processes.
+
+## Goals
+
+Creates a CRD for topics and a CRD for notifications, defining all the necessary and optional information for the various endpoints.
+
+Extends the rook operator to handle the CRs that would be submitted by users.
+
+## Implementation
+
+The CR for a topic configuration takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephBucketTopic
+metadata:
+  name: # name of the topic
+  namespace: # namespace where topic belongs
```

Comment:
Is this expected to be in the same namespace as rook and the object store (controlled by the admin)? Or can the application define their own notifications (in the same namespace as the app)?

Reply:
both are actually valid usecases, as the topic could be:
* part of the infrastructure, done by an admin, and provided as a service to the application developers
* used by the cluster admin itself (e.g. rgw auditing requirements)
* created for a specific application

can we allow for both (as I don't have info on how they are actually being used)?
if not, then since topics are not specific to buckets, they should probably be considered at the object store namespace

Reply:
The key here is that if the app needs to control the notifications, the properties must be on the BucketRequest, which is defined in the application namespace. The BucketNotification CR would also be created in the app namespace.

 If the admin controls the notifications, then they can be in the BucketClass parameters, and the BucketNotification CR would be created in the rook/object store namespace.

Reply:
For the best UX, the application or the namespace owner (e.g. a developer) must be able to do everything: create topic(s), configure notification(s) on bucket(s). If the user has to make a request and wait for an admin for that, it becomes easier to directly connect to the RGW and do it, which breaks the purpose of the simplified configuration we're trying to achieve here.
The fact that there may be "standard" notifications already created by admin that can be reused later on is the cherry on the cake, but not the goal. It's a simplifications for developers who will have their hands simultaneously in OpenShift, the endpoints like Kafka, their buckets,... So again, they have to be fully autonomous.

Reply:
I guess Bucket Notification CR need to created in Rook/object-store namespace. Because it is managed by Rook. And Bucket Request or OBC can be running in application namespace. I am not sure how we can tackle this. Do both need to be in the same namespace, or can rook manages objects out of its namespace? Can we link namespace of application and OBC/BR/BAR in the CRD of Bucket Notification than other way around?
@travisn @yuvalif ^^?

Reply:
@guimou ^^

Reply:
Rook can manage the CRs in other namespaces, it just needs the RBAC that will give the operator privileges to watch the CRs across namespaces. For example, by default Rook will watch for cluster CRs across all namespaces and can be restricted with [this setting](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/operator.yaml#L301).

I would also vote with @guimou that the main scenario for the notifications will be to be defined by the application in the app namespace, and then referenced by the OBC or BR which is defined in the same namespace.

Reply:
This is resolved, right? Topics and notifications are expected to be in the app namespace with the OBC?